gateway-service/src/main/java/org/zowe/apiml/gateway/zaas/ZaasExceptionHandler.java

```diff
@@ -0,0 +1,80 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.zowe.apiml.gateway.security.service.saf.SafIdtAuthException;
+import org.zowe.apiml.gateway.security.service.saf.SafIdtException;
+import org.zowe.apiml.gateway.security.service.schema.source.AuthSchemeException;
+import org.zowe.apiml.gateway.security.ticket.ApplicationNameNotFoundException;
+import org.zowe.apiml.message.api.ApiMessageView;
+import org.zowe.apiml.message.core.MessageService;
+import org.zowe.apiml.passticket.IRRPassTicketGenerationException;
+import org.zowe.apiml.security.common.token.TokenNotValidException;
+
+import javax.management.ServiceNotFoundException;
+
+@ControllerAdvice
+@RequiredArgsConstructor
+public class ZaasExceptionHandler {
+    private final MessageService messageService;
+
+    @ExceptionHandler(value = {IRRPassTicketGenerationException.class})
+    public ResponseEntity<ApiMessageView> handlePassTicketException(IRRPassTicketGenerationException ex) {
+        ApiMessageView messageView = messageService.createMessage("org.zowe.apiml.security.ticket.generateFailed",
+            ex.getErrorCode().getMessage()).mapToView();
+        return ResponseEntity
+            .status(ex.getHttpStatus())
+            .contentType(MediaType.APPLICATION_JSON)
+            .body(messageView);
+    }
+
+    @ExceptionHandler(value = {SafIdtException.class, SafIdtAuthException.class})
+    public ResponseEntity<ApiMessageView> handleSafIdtExceptions(RuntimeException ex) {
+        ApiMessageView messageView = messageService.createMessage("org.zowe.apiml.security.idt.failed", ex.getMessage()).mapToView();
+        return ResponseEntity
+            .status(HttpStatus.INTERNAL_SERVER_ERROR)
+            .contentType(MediaType.APPLICATION_JSON)
+            .body(messageView);
+    }
+
+    @ExceptionHandler(value = {ApplicationNameNotFoundException.class})
+    public ResponseEntity<ApiMessageView> handleApplIdNotFoundException(ApplicationNameNotFoundException ex) {
+        ApiMessageView messageView = messageService.createMessage("org.zowe.apiml.security.ticket.invalidApplicationName").mapToView();
+        return ResponseEntity
+            .status(HttpStatus.BAD_REQUEST)
+            .contentType(MediaType.APPLICATION_JSON)
+            .body(messageView);
+    }
+
+    @ExceptionHandler(value = {ServiceNotFoundException.class})
+    public ResponseEntity<ApiMessageView> handleServiceNotFoundException(ServiceNotFoundException ex) {
+        ApiMessageView messageView = messageService.createMessage("org.zowe.apiml.zaas.zosmf.noZosmfTokenReceived", ex.getMessage()).mapToView();
+        return ResponseEntity
+            .status(HttpStatus.SERVICE_UNAVAILABLE)
+            .contentType(MediaType.APPLICATION_JSON)
+            .body(messageView);
+    }
+
+    @ExceptionHandler(value = {TokenNotValidException.class, IllegalStateException.class, AuthSchemeException.class})
```

Comment:
Where does the tokenNotValidException happen? Isn't it rather 401? But it depends where it happens ...

Reply:
The TokenNotVAlidException could be for example from the parseOIDCToken() method when there is no mapping for the distributed ID.

Reply:
When there is no mapping, it is not a server error, right? But I am very tired today so I may be missing something.

Reply:
it may be considered as a server error since the token is valid but the server is not properly configured

Reply:
although the error message is somehow strange to me `zoweJwt.noToken`

Reply:
I would split it to 401. Invalid mapping is the invalid security setting of the user account, not the server.

Reply:
problem is that the user received 204 for authentication call, now 401 due to configuration error

Reply:
in the end, this is tiny detail and we are just wasting time

Reply:
TokenNotValidException, TokenExpiredException and AuthSchemeException (this is thrown in case of x509 mapping failure) are now generating 401 response. Then there is IllegalStateException (when z/OSMF is not configured) which is giving 500 response.